b2b_ecommerce/views.py

```diff
@@ -0,0 +1,137 @@
+    permission_classes = ()
+
+    def post(self, request, *args, **kwargs):  # pylint: disable=too-many-locals
+        """
+        Create a new unfulfilled Order from the user's basket
+        and return information used to submit to CyberSource.
+        """
+        try:
+            num_seats = request.data["num_seats"]
+            email = request.data["email"]
+            product_version_id = request.data["product_version_id"]
+        except KeyError as ex:
+            raise ValidationError(f"Missing parameter {ex.args[0]}")
+
+        try:
+            num_seats = int(num_seats)
+        except ValueError:
+            raise ValidationError("num_seats must be a number")
+
+        product_version = get_object_or_404(ProductVersion, id=product_version_id)
+        total_price = product_version.price * num_seats
+
+        base_url = request.build_absolute_uri("/")
+        order = B2BOrder.objects.create(
+            num_seats=num_seats,
+            email=email,
+            product_version=product_version,
+            total_price=total_price,
+            per_item_price=total_price,
+        )
+
+        receipt_url = urljoin(base_url, reverse("bulk-enrollment-code-receipt"))
+        if total_price == 0:
+            # If price is $0, don't bother going to CyberSource, just mark as fulfilled
+            order.status = B2BOrder.FULFILLED
+            order.save_and_log(None)
+
+            complete_b2b_order(order)
+
+            # This redirects the user to our order success page
+            payload = {}
+            url = receipt_url
+            method = "GET"
+        else:
+            # This generates a signed payload which is submitted as an HTML form to CyberSource
+            payload = generate_b2b_cybersource_sa_payload(
+                order=order, receipt_url=receipt_url, cancel_url=base_url
+            )
+            url = settings.CYBERSOURCE_SECURE_ACCEPTANCE_URL
+            method = "POST"
+
+        return Response({"payload": payload, "url": url, "method": method})
+
+
+class B2BOrderFulfillmentView(APIView):
+    """
+    View for order fulfillment API. This API is special in that only CyberSource should talk to it.
+    Instead of authenticating with OAuth or via session this looks at the signature of the message
+    to verify authenticity.
+    """
+
+    authentication_classes = ()
+    permission_classes = (IsSignedByCyberSource,)
+
+    def post(self, request, *args, **kwargs):
+        """
+        Confirmation from CyberSource which fulfills an existing Order.
+        """
+        # First, save this information in a receipt
+        receipt = B2BReceipt.objects.create(data=request.data)
+
+        # Link the order with the receipt if we can parse it
+        reference_number = request.data["req_reference_number"]
+        order = B2BOrder.objects.get_by_reference_number(reference_number)
+        receipt.order = order
+        receipt.save()
+
+        decision = request.data["decision"]
```

Comment:
Take or leave: This is another method that is almost identical to the non-bulk version. To reduce the risk of that, maybe this block from 109-128 can be moved into a separate API method called `determine_order_status(order, decision)` et. al. which would return the status string

Reply:
I tried to go in this direction but it's tricky because of the `return Response(status=200)`

Reply:
Seems like it's functionally equivalent if you return `None` in that function under those conditions, then in the `post` function just `return Response(status=200)` if the status is `None`. The function could be called `determine_order_status_change` to make it clearer. If there's no status change, there's no need to do anything else with this order, so you can just return from there.

Reply:
Ah I see. I just made the change